enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,677 @@
+---
+title: aws-custom-edge-machineset-local-zones
+authors:
+  - "@mtulio"
+reviewers:
+  - "@rvanderp3"
+  - "@faermanj"
+  - "@patrickdillon"
+  - "@elmiko"
+  - "@JoelSpeed"
+  - "@sdodson"
+approvers:
+  - "@patrickdillon"
+  - "@JoelSpeed"
+  - "@sdodson"
+api-approvers:
+  - None
+creation-date: 2022-09-05
+last-updated: 2022-09-05
+tracking-link:
+  - https://issues.redhat.com/browse/RFE-2782
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Allow Customer-Provisioned edge compute pool for AWS Local Zones
+
+## Summary
+
+The edge adoption is increasing and providers are launching many offerings that fill
+that area. AWS launched the [Local Zones](https://aws.amazon.com/about-aws/global-infrastructure/localzones/)
+and the [Wavelength](https://aws.amazon.com/wavelength/features/) zones to extend the
+VPC network to the metropolitan area from big centers, and for Radio Access Network,
+respectively, to allow application developers to deliver low-latency applications
+at scale to the end users.
+
+As the resources in that locations were designed for specific workloads, and
+also have limited and more expensive cloud services, cluster operators should be
+able to create worker nodes on the edge networks on day zero operations.
+
+This enhancement proposes a new compute pool named `edge`, which will create
+Machine Sets for each availability zone on the AWS Local Zones to provision compute
+resources with `NoSchedule` taint.
+
+The EP is divided into three phases:
+
+- Phase 0: Research. Understand the Local Zones using the OKD/OCP, create use case, and
+  provides the documentation of the installation of a cluster in an existing VPC.
+- Phase 1: Install a cluster in user provided network introducing to the installer a new
+  compute pool named `edge`. The MachineSet manifests will be created for each subnet
+  provided on `platform.aws.subnets` for Local Zones. The Machine will be protected with
+  taints to avoid running the normal platform workloads.
+- Phase 2: The installer should automate the network resource creation when the user
+  provides zone names with the type `local-zone` on the `platform.aws.zones` or
+  `compute[?name=="edge"].platform.aws.zones`.
+
+Phases 1 and 2 are covered by this enhancement.
+
+## Motivation
+
+Presently, the installer does not support Local Zone subnets on day zero operations
```

Comment:
> I'm interested to know how hard it is to set up these edge nodes as a day 2 operation, what's involved?

The steps are like this:
- check if the VPC has enough CIDR blocks to add new subnets
- opt-in zone group
- create subnets on the Local Zone, setting the Tags to avoid the ingress controller discovery of this subnet
- associate the subnet with a public route table
- check the instance type supported on the AZ
- create the MachineSet for each Local Zone AZ, setting the taints for that MachineSet (for reasons presented on this EP on why not sharing the nodes on the edge with normal workloads)
- scale the machines

If the customer would like to deploy in the private subnet, it could become more complex as it needs to use the Nat GW from the parent region.

If the VPC haven't enough blocks, the customer will not be able to use Local Zones.

> Where has the need for day 0 come from? Could it be/would it be simpler to have day 2 set up only?

[RFE-2782](https://issues.redhat.com/browse/RFE-2782). Mainly telecom customers may be able to fully deploy workloads on the edge without extra effort or hacking on Day-2, or Day-0 adding extra manifests or installing in existing VPC. The second phase will implement support to installer create all required resources.
cc @makentenza could you please add your points of customer requesting that feature?

Reply:
Both the "what does this look like today" flow and the motivation that we have customers looking for 1 touch installations across many edges are useful contexts that would be good to capture in the enhancement IMO

Reply:
@JoelSpeed thanks for raising it. The ["Motivation"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R60-R97) has been updated with your feedback on this thread.